structured/privilege.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package structured
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/security"
+)
+
+// PrivilegeList is a list of privileges.
+type PrivilegeList []Privilege
+
+func (pl PrivilegeList) Len() int {
+	return len(pl)
+}
+
+func (pl PrivilegeList) Swap(i, j int) {
+	pl[i], pl[j] = pl[j], pl[i]
+}
+
+func (pl PrivilegeList) Less(i, j int) bool {
+	return pl[i] < pl[j]
+}
+
+// privilegeValues is a convenience variable holding a sorted
+// list of privilege values.
+var privilegeValues = func() PrivilegeList {
+	pv := make(PrivilegeList, len(Privilege_name), len(Privilege_name))
+	i := 0
+	for v := range Privilege_name {
+		pv[i] = Privilege(v)
+		i++
+	}
+	sort.Sort(pv)
+	return pv
+}()
+
+// String implements the Stringer interface.
+// This keeps the existing order and uses ", " as separator.
+func (pl PrivilegeList) String() string {
+	ret := make([]string, len(pl), len(pl))
+	for i, p := range pl {
+		ret[i] = Privilege_name[int32(p)]
```

Comment:
oh. yeah. I actually takes advantage of that in the error in sql/descriptor.go:
`fmt.Errorf("user %s does not have %s privilege on %s %s",
    p.user, privilege, descriptor.TypeName(), descriptor.GetName())`

Fixing.